src/libertem/analysis/helper.py

```diff
@@ -0,0 +1,112 @@
+from libertem.web.notebook_generator.code_template import TemplateBase
+
+
+class GeneratorHelper(TemplateBase):
+
+    short_name = None
+    api = None
+
+    def __init__(self, params):
+        self.params = params
+
+    def get_dependency(self):
+        """
+        Get analysis dependencies.
+        """
+        return None
+
+    def convert_params(self):
+        """
+        Format analysis parameters.
+        """
+        return None
+
+    def get_plot(self):
+        """
+        Get code for ploting analysis.
+        """
+        return None
+
+    def get_docs(self):
+        """
+        Get documentation for analysis.
+        """
+        return None
+
+    def format_docs(self, docs_rst):
+        """
+        function to convert RST to MD format
+        """
+        output = convert_text(docs_rst, 'commonmark', format='rst')
+        # converting heading level
+        output = output.replace('#', '###')
+        return output
+
+    def get_roi_code(self):
+
+        if 'roi' in self.params.keys():
+            data = {'roi_params': self.params['roi']}
+            roi = self.format_template(self.temp_roi, data)
+        else:
+            roi = f"roi = {self.short_name}_analysis.get_roi()"
+
+        return roi
+
+    def get_analysis(self):
+
+        params_ = self.convert_params()
+        roi = self.get_roi_code()
+
+        data = {'short': self.short_name,
+                'analysis_api': self.api,
+                'params': params_,
+                'roi': roi}
+
+        analy_ = self.format_template(self.temp_analysis, data)
+
+        return analy_
+
+
+# can be moved to some other place or any other idea!
+temp_cluster_controller = """
+stddev_udf = StdDevUDF()
+roi = cluster_analysis.get_sd_roi()
+sd_udf_results = ctx.run_udf(dataset=ds, udf=stddev_udf, roi=roi, progress=True)
+sd_udf_results = consolidate_result(sd_udf_results)
+
+
+center = (parameters["cy"], parameters["cx"])
+rad_in = parameters["ri"]
+rad_out = parameters["ro"]
+n_peaks = parameters["n_peaks"]
+min_dist = parameters["min_dist"]
+sstd = sd_udf_results['std']
+sshape = sstd.shape
+if not (center is None or rad_in is None or rad_out is None):
+    mask_out = 1*_make_circular_mask(center[1], center[0], sshape[1], sshape[0], rad_out)
+    mask_in = 1*_make_circular_mask(center[1], center[0], sshape[1], sshape[0], rad_in)
+    mask = mask_out - mask_in
+    masked_sstd = sstd*mask
+else:
+    masked_sstd = sstd
+
+coordinates = peak_local_max(masked_sstd, num_peaks=n_peaks, min_distance=min_dist)
+
+y = coordinates[..., 0]
+x = coordinates[..., 1]
+z = range(len(y))
+
+mask = sparse.COO(
+    shape=(len(y), ) + tuple(ds.shape.sig),
+    coords=(z, y, x), data=1
+)
+
+udf = ApplyMasksUDF(
+    mask_factories=lambda: mask, mask_count=len(y), mask_dtype=np.uint8,
+    use_sparse=True
+)
+
+udf_results = ctx.run_udf(dataset=ds, udf=udf, progress=True)
+cluster_result = cluster_analysis.get_udf_results(udf_results=udf_results, roi=roi)
+"""
```

Comment:
Why is this here? It looks like something specific to the cluster analysis? In that case it should be close to there and not in a general part of the code.

Generally, as already mentioned above, would it perhaps be possible to refactor this "temp_cluster_controller" and the actual controller in clust.py in such a way that they share most of the code and that the source code is pulled from the actual implementation by introspection instead of duplicating it?